src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -282,6 +306,10 @@ void GcsNodeManager::OnNodeFailure(const NodeID &node_id,
   if (auto node = RemoveNode(node_id, /* is_intended = */ false)) {
     node->set_state(rpc::GcsNodeInfo::DEAD);
     node->set_end_time_ms(current_sys_time_ms());
+    if (node->death_info().reason() == rpc::NodeDeathInfo::UNSPECIFIED) {
```

Comment:
Add SetNodePreempted

and add unit test

`# SetNodePreempted -> OnNodeFailure -> verify reason is correctly set`